### PR TITLE
Addons: Support arbitrary additional objects

### PR DESCRIPTION
**Pull request by justinsb**
We will be managing cluster addons using CRDs, and so we want to be
able to apply arbitrary objects as part of cluster bringup.

Start by allowing (behind a feature-flag) for arbitrary objects to be
specified.

### CODE REVIEW

**Comment by justinsb** [comment]
No longer WIP - and I sent #9374 that shows how this can be used

File: upup/pkg/fi/cloudup/bootstrapchannelbuilder.go [hunk 1]
```diff
@@ -95,7 +97,50 @@ func (b *BootstrapChannelBuilder) Build(c *fi.ModelBuilderContext) error {
...
+		location := key + "/default.yaml"
+
+		a := &channelsapi.AddonSpec{
+			Name:     fi.String(key),
+			Version:  fi.String(version),
+			Selector: map[string]string{"k8s-addon": key},
+			Manifest: fi.String(location),
+		}
+
+		if a.Id != "" {
```
Comment (johngmyers) [inline upup/pkg/fi/cloudup/bootstrapchannelbuilder.go:114]: How can this ever be true?

File: upup/pkg/fi/cloudup/bootstrapchannelbuilder.go [hunk 2]
```diff
@@ -95,7 +97,50 @@ func (b *BootstrapChannelBuilder) Build(c *fi.ModelBuilderContext) error {
...
+
+		rawManifest := string(manifestBytes)
+
+		manifestHash, err := utils.HashString(rawManifest)
+		if err != nil {
+			return fmt.Errorf("error hashing manifest: %v", err)
+		}
+		a.ManifestHash = manifestHash
+
+		tasks[name] = &fitasks.ManagedFile{
```
Comment (johngmyers) [inline upup/pkg/fi/cloudup/bootstrapchannelbuilder.go:136]: Why not call `c.AddTask()`? This seems an indirect, hacky way of adding a task.

File: upup/pkg/fi/cloudup/bootstrapchannelbuilder.go (hunk 1 above)
Comment (justinsb) [reply upup/pkg/fi/cloudup/bootstrapchannelbuilder.go]: Good call - removed (I think I was aiming for parity with the existing code, but I don't know)

File: upup/pkg/fi/cloudup/bootstrapchannelbuilder.go (hunk 2 above)
Comment (justinsb) [reply upup/pkg/fi/cloudup/bootstrapchannelbuilder.go]: Agreed, and I don't know why.  Changed to use `AddTask`

**Comment by justinsb** [comment]
Rebased and fixed up the tests that I had broken.

**Comment by justinsb** [comment]
Absolutely agree on the need for some form of update/replace logic @johngmyers 

For why they aren't (currently) granular, it's because I am guessing about how this would be used (i.e. how I will use it).  I imagine that I'm managing my cluster definitions in git, and so there's a directory tree called `addons` alongside my cluster.yaml.   So for this case, the "big bag of objects" model is probably all I want.

But this is part of why this is feature flagged - we need to get feedback on whether people want to manage them more granularly.  It would be easy enough to do (and easy enough to change to it, as well, I think).  I just want to avoid it if it would be more confusing.  We also don't know whether people will just reference some operators here, or whether they will expand out the raw objects (i.e. will there be 5 objects or 100?)

I am imagining templating as something that happens before kops, creating the addons directory.  That lets people use whatever tooling they want to (including their own generation scripts, which I imagine will be pretty common for advanced use cases)

**Pull request merged**